File: packages/slides/src/qc.ts
import type { DeckSpec, SlideKind, SlideSpec } from './types'

// Deterministic (no LLM) layout QC for DeckSpec. DeckView.tsx is a fixed
// HTML/flex template per SlideKind with no absolute x/y/w/h — CSS
// container-query units (cqw) sized relative to the slide's own width,
// flowing top-to-bottom, clipped by `overflow:hidden` on the 6cqw-inset
// content box (SlideView).
//
// Because there is no real box geometry to compare, this estimates it: a
// character-width text-metrics heuristic (proportional sans-serif average
// glyph width ≈ 0.52× font-size) turns each text field's length + the
// template's known max-width into an estimated wrapped line count, and sums
// per-kind stacked element heights (title/eyebrow/subtitle/body/bullets, plus
// each kind's fixed gaps) against the 44.25cqw content-height budget that
// falls out of the 16:9 stage minus its 6cqw margins. That sum crossing the
// budget is this model's "out of bounds" — SlideView clips it, so overflow
// really does mean lost content, not just visual crowding. A genuine
// *overlap* check (two independent boxes intersecting) isn't meaningful for
// a single-column flow layout — the one real overlap risk here is stacked
// content growing tall enough to run into a kind's FIXED-position decoration
// (the accent bar under a title/closing slide, the page-number chip in a
// slide's bottom-right corner), so that risk is checked explicitly instead of
// faked as a generic AABB test. All thresholds are heuristic estimates, not
// pixel-exact — this is a coarse "is this dangerously long" filter, not a
// layout engine.
//
// When the OOXML canvas model lands (real x/y/w/h shapes — see
// go/pptxpatch), a second, exact overlap/overflow checker over THAT geometry
// can use AABB intersection; this checker stays as DeckSpec's own
// pre-compile QC pass.

export type QCIssueKind = 'overflow' | 'overlap' | 'out-of-bounds'

export interface QCIssue {
  slideId: string
  slideIndex: number
  kind: QCIssueKind
  /** Human-readable, meant to be fed straight back to a revise/fix prompt. */
  message: string
}

// ---- Stage geometry, all in "cqw" (% of slide width) — mirrors DeckView.tsx ----

const STAGE_W = 100
const STAGE_H = (STAGE_W * 9) / 16 // 16:9 aspect, height in the same cqw unit
const MARGIN = 6 // SlideView's inset on all 4 sides
const CONTENT_W = STAGE_W - 2 * MARGIN
const CONTENT_H = STAGE_H - 2 * MARGIN // ≈ 44.25

/** Reserved bottom strip: the "N / total" page-number chip every non-title
 *  kind renders at right/bottom ~2.4cqw — content stacking into this band
 *  risks visually colliding with it. */
const PAGE_NUMBER_RESERVE_H = 4
/** Reserved bottom strip for the accent bar under title/closing slides. */
const ACCENT_BAR_RESERVE_H = 3

/** Average glyph width as a fraction of font-size, for a proportional
 *  display/body sans stack — a standard typographic rule of thumb (not
 *  measured per font; this is an estimate, documented as such throughout). */
const AVG_CHAR_WIDTH = 0.52

/** Strip the bold/italic emphasis markers DeckView's renderInline consumes —
 *  they don't render as glyphs, so they shouldn't count toward width. */
function plainLength(text: string): number {
  return text.replace(/\*\*([^*]+)\*\*/g, '$1').replace(/\*([^*]+)\*/g, '$1').length
}

/** Estimated wrapped line count for one paragraph of text at fontSizeCqw,
 *  inside a box boxWidthCqw wide. Explicit '\n' segments (pre-wrap fields
 *  like body/notes) each wrap independently. Empty text = 0 lines. */
export function estimateWrappedLines(text: string, boxWidthCqw: number, fontSizeCqw: number): number {
  if (!text) return 0
  const charsPerLine = Math.max(1, Math.floor(boxWidthCqw / (fontSizeCqw * AVG_CHAR_WIDTH)))
  let lines = 0
  for (const segment of text.split('\n')) {
    const len = plainLength(segment)
    lines += len === 0 ? 1 : Math.ceil(len / charsPerLine)
  }
  return lines
}

/** Estimated rendered height (cqw) of one paragraph: lines × fontSize × lineHeight. */
function textBlockHeight(text: string, boxWidthCqw: number, fontSizeCqw: number, lineHeight: number): number {
  return estimateWrappedLines(text, boxWidthCqw, fontSizeCqw) * fontSizeCqw * lineHeight
}

/** Bulleted list height: each item wraps independently within boxWidthCqw
 *  (minus the list's left padding), items separated by the template's gap. */
function bulletsHeight(items: string[] | undefined, boxWidthCqw: number): number {
  const list = items ?? []
  if (list.length === 0) return 0
  const fontSizeCqw = 2.6
  const gap = 1.5
  const innerW = boxWidthCqw - 2.4 // ul padding-left
  const linesTotal = list.reduce((sum, item) => sum + Math.max(1, estimateWrappedLines(item, innerW, fontSizeCqw)), 0)
  return linesTotal * fontSizeCqw * 1.55 + gap * (list.length - 1)
}

function eyebrowHeight(eyebrow: string | undefined): number {
  return eyebrow ? 1.4 * 1.2 + 1.6 : 0 // font-size*lineHeight-ish + its own margin-bottom
}

interface KindBudget {
  /** Estimated stacked content height (cqw) for this slide. */
  contentH: number
  /** Bottom reserve this kind's fixed decoration needs (cqw), if any. */
  reserve: number
  reserveLabel: string
  /** Per-field notes for width-driven issues worth calling out individually. */
  fieldNotes: string[]
}

function budgetFor(slide: SlideSpec): KindBudget {
  const notes: string[] = []
  switch (slide.kind) {
    case 'title': {
      const titleW = CONTENT_W * 0.8
      const subW = CONTENT_W * 0.62
      const h =
        eyebrowHeight(slide.eyebrow) +
        textBlockHeight(slide.title ?? '', titleW, 7, 1.05) +
        (slide.subtitle ? textBlockHeight(slide.subtitle, subW, 2.6, 1.55) + 2.4 : 0) +
        (slide.body ? textBlockHeight(slide.body, subW, 2, 1.55) + 1.6 : 0)
      return { contentH: h, reserve: ACCENT_BAR_RESERVE_H, reserveLabel: 'the title bar accent', fieldNotes: notes }
    }
    case 'section': {
      const titleW = CONTENT_W * 0.78
      const subW = CONTENT_W * 0.64
      const h =
        eyebrowHeight(slide.eyebrow) +
        textBlockHeight(slide.title ?? '', titleW, 5.4, 1.05) +
        (slide.subtitle || slide.body ? textBlockHeight(slide.subtitle ?? slide.body ?? '', subW, 2.3, 1.55) + 2 : 0)
      return { contentH: h, reserve: PAGE_NUMBER_RESERVE_H, reserveLabel: 'the page-number chip', fieldNotes: notes }
    }
    case 'quote': {
      const quoteW = CONTENT_W * 0.8 - 5 // minus the callout box's own horizontal padding (5cqw)
      const h =
        8 + // callout vertical padding (4cqw top + 4cqw bottom)
        textBlockHeight(slide.quote ?? '', quoteW, 3.4, 1.4) +
        (slide.subtitle ? textBlockHeight(slide.subtitle, quoteW, 1.6, 1.4) + 2.2 : 0)
      return { contentH: h, reserve: PAGE_NUMBER_RESERVE_H, reserveLabel: 'the page-number chip', fieldNotes: notes }
    }
    case 'two-col': {
      const colW = (CONTENT_W - 5) / 2
      const leftH = bulletsHeight(slide.bullets, colW)
      const rightH = bulletsHeight(slide.bulletsRight, colW)
      if (Math.abs(leftH - rightH) > 12) {
        notes.push(
          `two-col columns are visually unbalanced (left ≈${leftH.toFixed(1)}cqw tall vs right ≈${rightH.toFixed(1)}cqw) — consider redistributing items`,
        )
      }
      const colH = Math.max(leftH, rightH)
      const h = eyebrowHeight(slide.eyebrow) + textBlockHeight(slide.title ?? '', CONTENT_W, 4, 1.05) + 3 + colH
      return { contentH: h, reserve: PAGE_NUMBER_RESERVE_H, reserveLabel: 'the page-number chip', fieldNotes: notes }
    }
    case 'closing': {
      const h =
        textBlockHeight(slide.title ?? 'Thank you', CONTENT_W, 5.6, 1.05) +
        (slide.subtitle ? textBlockHeight(slide.subtitle, CONTENT_W, 2.2, 1.55) + 2 : 0)
      return { contentH: h, reserve: ACCENT_BAR_RESERVE_H + 3.4, reserveLabel: 'the closing accent bar', fieldNotes: notes }
    }
    case 'bullets':
    default: {
      const bodyW = CONTENT_W * 0.8
      const h =
        eyebrowHeight(slide.eyebrow) +
        (slide.title ? textBlockHeight(slide.title, CONTENT_W, 4, 1.05) + 3 : 0) +
        (slide.body ? textBlockHeight(slide.body, bodyW, 2.1, 1.55) + 2.4 : 0) +
        bulletsHeight(slide.bullets, CONTENT_W)
      return { contentH: h, reserve: PAGE_NUMBER_RESERVE_H, reserveLabel: 'the page-number chip', fieldNotes: notes }
    }
  }
}

function fmt(n: number): string {
  return n.toFixed(1)
}

/** Audit one slide. Pure function — same input, same output, no rendering. */
export function auditSlide(slide: SlideSpec, slideIndex: number): QCIssue[] {
  const issues: QCIssue[] = []
  const budget = budgetFor(slide)
  const push = (kind: QCIssueKind, message: string) => issues.push({ slideId: slide.id, slideIndex, kind, message })

  for (const note of budget.fieldNotes) push('overlap', note)

  const available = CONTENT_H
  if (budget.contentH > available) {
    push(
      'out-of-bounds',
      `content is estimated at ≈${fmt(budget.contentH)}cqw tall, ${fmt(budget.contentH - available)}cqw past the ${fmt(available)}cqw slide content area — it will be clipped (shorten text, split into another slide, or reduce bullet count)`,
    )
  } else if (budget.reserve > 0 && budget.contentH > available - budget.reserve) {
    push(
      'overlap',
      `content is estimated at ≈${fmt(budget.contentH)}cqw tall, within ${fmt(available - budget.contentH)}cqw of ${budget.reserveLabel} — it may visually collide with it`,
    )
  }

  return issues
}

/** Audit every slide in a deck. Empty array = clean deck. */
export function auditDeck(deck: DeckSpec): QCIssue[] {
  const issues: QCIssue[] = []
  deck.slides.forEach((slide, i) => issues.push(...auditSlide(slide, i)))
  return issues
}

/** Format issues as text suitable for feeding back into an AI revise prompt. */
export function formatDeckAudit(issues: QCIssue[]): string {
  if (issues.length === 0) return '<deck-audit>Passed: no estimated overflow/overlap issues.</deck-audit>'
  const body = issues.map((iss) => `- Slide ${iss.slideIndex + 1} (${iss.slideId}) [${iss.kind}]: ${iss.message}`).join('\n')
  return `<deck-audit>Found ${issues.length} issue(s):\n${body}\n</deck-audit>`
}

export type { SlideKind }

File: go/slidesqc/qc.go
package slidesqc

import (
	"fmt"
	"math"
	"regexp"
	"strings"
)

// ---- Stage geometry, all in "cqw" (% of slide width) — mirrors DeckView.tsx
// and MUST stay numerically identical to qc.ts. See that file for the full
// rationale: there is no real box geometry in the HTML-first renderer this
// audits, so overflow/overlap are estimated from a character-width text
// heuristic against each slide kind's known stacked-element layout, not
// measured from real boxes.

// NOTE: these are typed float64 constants, not untyped integer constants —
// stageH's division must be true (56.25), not truncating integer division,
// to match qc.ts' floating-point arithmetic exactly.
const (
	stageW   float64 = 100
	stageH   float64 = (stageW * 9) / 16 // 16:9 aspect, height in the same cqw unit
	margin   float64 = 6                 // SlideView's inset on all 4 sides
	contentW float64 = stageW - 2*margin
	contentH float64 = stageH - 2*margin // ≈ 44.25
)

// pageNumberReserveH is the reserved bottom strip for the "N / total"
// page-number chip every non-title kind renders — content stacking into
// this band risks visually colliding with it.
const pageNumberReserveH = 4

// accentBarReserveH is the reserved bottom strip for the accent bar under
// title/closing slides.
const accentBarReserveH = 3

// avgCharWidth is the average glyph width as a fraction of font-size, for a
// proportional display/body sans stack — a standard typographic rule of
// thumb (not measured per font; an estimate, as throughout this package).
const avgCharWidth = 0.52

var (
	boldRe   = regexp.MustCompile(`\*\*([^*]+)\*\*`)
	italicRe = regexp.MustCompile(`\*([^*]+)\*`)
)

// plainLength strips the bold/italic emphasis markers DeckView's renderInline
// consumes — they don't render as glyphs, so they shouldn't count toward
// width. Counted in runes (not bytes) to match JS string .length semantics
// closely enough for this heuristic's purposes.
func plainLength(text string) int {
	stripped := italicRe.ReplaceAllString(boldRe.ReplaceAllString(text, "$1"), "$1")
	return len([]rune(stripped))
}

// EstimateWrappedLines returns the estimated wrapped line count for one
// paragraph of text at fontSizeCqw, inside a box boxWidthCqw wide. Explicit
// '\n' segments (pre-wrap fields like body/notes) each wrap independently.
// Empty text = 0 lines.
func EstimateWrappedLines(text string, boxWidthCqw, fontSizeCqw float64) int {
	if text == "" {
		return 0
	}
	charsPerLine := int(math.Max(1, math.Floor(boxWidthCqw/(fontSizeCqw*avgCharWidth))))
	lines := 0
	for _, segment := range strings.Split(text, "\n") {
		l := plainLength(segment)
		if l == 0 {
			lines++
		} else {
			lines += int(math.Ceil(float64(l) / float64(charsPerLine)))
		}
	}
	return lines
}

// textBlockHeight is the estimated rendered height (cqw) of one paragraph:
// lines × fontSize × lineHeight.
func textBlockHeight(text string, boxWidthCqw, fontSizeCqw, lineHeight float64) float64 {
	return float64(EstimateWrappedLines(text, boxWidthCqw, fontSizeCqw)) * fontSizeCqw * lineHeight
}

// bulletsHeight is the bulleted-list height: each item wraps independently
// within boxWidthCqw (minus the list's left padding), items separated by the
// template's gap.
func bulletsHeight(items []string, boxWidthCqw float64) float64 {
	if len(items) == 0 {
		return 0
	}
	const fontSizeCqw = 2.6
	const gap = 1.5
	innerW := boxWidthCqw - 2.4 // ul padding-left
	linesTotal := 0.0
	for _, item := range items {
		linesTotal += math.Max(1, float64(EstimateWrappedLines(item, innerW, fontSizeCqw)))
	}
	return linesTotal*fontSizeCqw*1.55 + gap*float64(len(items)-1)
}

func eyebrowHeight(eyebrow string) float64 {
	if eyebrow == "" {
		return 0
	}
	return 1.4*1.2 + 1.6 // font-size*lineHeight-ish + its own margin-bottom
}

// kindBudget is the estimated layout budget for one slide.
type kindBudget struct {
	// contentH is the estimated stacked content height (cqw) for this slide.
	contentH float64
	// reserve is the bottom reserve this kind's fixed decoration needs
	// (cqw), if any.
	reserve      float64
	reserveLabel string
	// fieldNotes are per-field notes for width-driven issues worth calling
	// out individually (e.g. unbalanced two-col columns).
	fieldNotes []string
}

func budgetFor(slide SlideSpec) kindBudget {
	var notes []string
	switch slide.Kind {
	case KindTitle:
		titleW := contentW * 0.8
		subW := contentW * 0.62
		h := eyebrowHeight(slide.Eyebrow) + textBlockHeight(slide.Title, titleW, 7, 1.05)
		if slide.Subtitle != "" {
			h += textBlockHeight(slide.Subtitle, subW, 2.6, 1.55) + 2.4
		}
		if slide.Body != "" {
			h += textBlockHeight(slide.Body, subW, 2, 1.55) + 1.6
		}
		return kindBudget{contentH: h, reserve: accentBarReserveH, reserveLabel: "the title bar accent", fieldNotes: notes}

	case KindSection:
		titleW := contentW * 0.78
		subW := contentW * 0.64
		h := eyebrowHeight(slide.Eyebrow) + textBlockHeight(slide.Title, titleW, 5.4, 1.05)
		if slide.Subtitle != "" || slide.Body != "" {
			text := slide.Subtitle
			if text == "" {
				text = slide.Body
			}
			h += textBlockHeight(text, subW, 2.3, 1.55) + 2
		}
		return kindBudget{contentH: h, reserve: pageNumberReserveH, reserveLabel: "the page-number chip", fieldNotes: notes}

	case KindQuote:
		quoteW := contentW*0.8 - 5 // minus the callout box's own horizontal padding (5cqw)
		h := 8.0 + textBlockHeight(slide.Quote, quoteW, 3.4, 1.4)
		if slide.Subtitle != "" {
			h += textBlockHeight(slide.Subtitle, quoteW, 1.6, 1.4) + 2.2
		}
		return kindBudget{contentH: h, reserve: pageNumberReserveH, reserveLabel: "the page-number chip", fieldNotes: notes}

	case KindTwoCol:
		colW := (contentW - 5) / 2
		leftH := bulletsHeight(slide.Bullets, colW)
		rightH := bulletsHeight(slide.BulletsRight, colW)
		if math.Abs(leftH-rightH) > 12 {
			notes = append(notes, fmt.Sprintf(
				"two-col columns are visually unbalanced (left ≈%scqw tall vs right ≈%scqw) — consider redistributing items",
				fmtNum(leftH), fmtNum(rightH),
			))
		}
		colH := math.Max(leftH, rightH)
		h := eyebrowHeight(slide.Eyebrow) + textBlockHeight(slide.Title, contentW, 4, 1.05) + 3 + colH
		return kindBudget{contentH: h, reserve: pageNumberReserveH, reserveLabel: "the page-number chip", fieldNotes: notes}

	case KindClosing:
		title := slide.Title
		if title == "" {
			title = "Thank you"
		}
		h := textBlockHeight(title, contentW, 5.6, 1.05)
		if slide.Subtitle != "" {
			h += textBlockHeight(slide.Subtitle, contentW, 2.2, 1.55) + 2
		}
		return kindBudget{contentH: h, reserve: accentBarReserveH + 3.4, reserveLabel: "the closing accent bar", fieldNotes: notes}

	case KindBullets:
		fallthrough
	default:
		bodyW := contentW * 0.8
		h := eyebrowHeight(slide.Eyebrow)
		if slide.Title != "" {
			h += textBlockHeight(slide.Title, contentW, 4, 1.05) + 3
		}
		if slide.Body != "" {
			h += textBlockHeight(slide.Body, bodyW, 2.1, 1.55) + 2.4
		}
		h += bulletsHeight(slide.Bullets, contentW)
		return kindBudget{contentH: h, reserve: pageNumberReserveH, reserveLabel: "the page-number chip", fieldNotes: notes}
	}
}

// fmtNum matches JS's Number.prototype.toFixed(1) rounding behavior (round
// half AWAY from zero) rather than Go's fmt.Sprintf("%.1f", n), which rounds
// exact binary midpoints (e.g. 44.25) half TO EVEN (44.2, not 44.3) — a
// real mismatch this package hit in cross-checking against qc.ts, since
// several of these budgets land on exact quarter-cqw sums.
func fmtNum(n float64) string {
	rounded := math.Round(n*10) / 10
	return fmt.Sprintf("%.1f", rounded)
}

// AuditSlide audits one slide. Pure function — same input, same output, no
// rendering.
func AuditSlide(slide SlideSpec, slideIndex int) []QCIssue {
	var issues []QCIssue
	budget := budgetFor(slide)
	push := func(kind QCIssueKind, message string) {
		issues = append(issues, QCIssue{SlideID: slide.ID, SlideIndex: slideIndex, Kind: kind, Message: message})
	}

	for _, note := range budget.fieldNotes {
		push(IssueOverlap, note)
	}

	available := contentH
	switch {
	case budget.contentH > available:
		push(IssueOutOfBounds, fmt.Sprintf(
			"content is estimated at ≈%scqw tall, %scqw past the %scqw slide content area — it will be clipped (shorten text, split into another slide, or reduce bullet count)",
			fmtNum(budget.contentH), fmtNum(budget.contentH-available), fmtNum(available),
		))
	case budget.reserve > 0 && budget.contentH > available-budget.reserve:
		push(IssueOverlap, fmt.Sprintf(
			"content is estimated at ≈%scqw tall, within %scqw of %s — it may visually collide with it",
			fmtNum(budget.contentH), fmtNum(available-budget.contentH), budget.reserveLabel,
		))
	}

	return issues
}

// AuditDeck audits every slide in a deck. An empty (nil) slice = clean deck.
func AuditDeck(deck DeckSpec) []QCIssue {
	var issues []QCIssue
	for i, slide := range deck.Slides {
		issues = append(issues, AuditSlide(slide, i)...)
	}
	return issues
}

// FormatDeckAudit formats issues as text suitable for feeding back into an
// AI revise prompt (mirrors qc.ts' formatDeckAudit: a compact block a model
// can act on).
func FormatDeckAudit(issues []QCIssue) string {
	if len(issues) == 0 {
		return "<deck-audit>Passed: no estimated overflow/overlap issues.</deck-audit>"
	}
	lines := make([]string, len(issues))
	for i, iss := range issues {
		lines[i] = fmt.Sprintf("- Slide %d (%s) [%s]: %s", iss.SlideIndex+1, iss.SlideID, iss.Kind, iss.Message)
	}
	return fmt.Sprintf("<deck-audit>Found %d issue(s):\n%s\n</deck-audit>", len(issues), strings.Join(lines, "\n"))
}
